pkg/reduce/readloop/readloop.go

```diff
@@ -125,67 +124,78 @@ func (rl *ReadLoop) Process(ctx context.Context, messages []*isb.ReadMessage) {
 		Jitter:   0.1,
 	}
 
-	for _, m := range messages {
-		// identify and add window for the message
-		var ctxClosedErr error
-		windows := rl.upsertWindowsAndKeys(m)
-		// for each window we will have a PBQ. A message could belong to multiple windows (e.g., sliding).
-		for _, kw := range windows {
-			// identify partition for message
-			partitionID := partition.ID{
-				Start: kw.StartTime(),
-				End:   kw.EndTime(),
-				Key:   m.Key,
-			}
-
-			q := rl.associatePBQAndPnF(ctx, partitionID)
+	select {
+	case <-ctx.Done():
+		return
+
+	default:
+		for _, m := range messages {
+			// identify and add window for the message
+			windows := rl.upsertWindowsAndKeys(m)
+			// for each window we will have a PBQ. A message could belong to multiple windows (e.g., sliding).
+			for _, kw := range windows {
+				// identify partition for message
+				partitionID := partition.ID{
+					Start: kw.StartTime(),
+					End:   kw.EndTime(),
+					Key:   m.Key,
+				}
 
-			// write the message to PBQ
-			attempt := 0
-			ctxClosedErr = wait.ExponentialBackoffWithContext(ctx, pbqWriteBackoff, func() (done bool, err error) {
-				rErr := q.Write(ctx, m)
-				attempt += 1
-				if rErr != nil {
-					rl.log.Errorw("Failed to write message", zap.Any("msgOffSet", m.ReadOffset.String()), zap.String("partitionID", partitionID.String()), zap.Any("attempt", attempt), zap.Error(rErr))
-					return false, nil
+				q := rl.associatePBQAndPnF(ctx, partitionID)
+
+				// write the message to PBQ
+				// we use a background context since these messages have to be written to pbq
+				cctx := context.Background()
+				attempt := 0
+				timeoutErr := wait.ExponentialBackoff(pbqWriteBackoff, func() (done bool, err error) {
```

Comment:
Is this going to be timeout shortly? The step number is set to `math.MaxInt`, this means, there's no way to shutdown until the pod is forced to terminate after grace shutdown period.